src/librustc_lint/unused.rs

```diff
@@ -179,14 +235,45 @@ impl<'a, 'tcx> LateLintPass<'a, 'tcx> for UnusedResults {
                         vec![]
                     };
                     for (i, ty) in tys.iter().map(|k| k.expect_ty()).enumerate() {
-                        let descr_post_path = &format!(" in tuple element {}", i);
+                        let descr_post = &format!(" in tuple element {}", i);
                         let span = *spans.get(i).unwrap_or(&span);
-                        if check_must_use_ty(cx, ty, expr, span, descr_post_path) {
-                            has_emitted = true;
-                        }
+                        has_emitted |= check_must_use_ty(
+                            cx, ty, expr, span, descr_pre, descr_post, plural);
                     }
                     has_emitted
                 }
+                ty::Array(ty, mut len) => {
+                    // Try to evaluate the length if it's unevaluated.
+                    if let ConstValue::Unevaluated(def_id, substs) = len.val {
+                        let instance = ty::Instance::resolve(
+                            cx.tcx.global_tcx(),
+                            cx.param_env,
+                            def_id,
+                            substs,
+                        ).unwrap();
+                        let global_id = GlobalId {
+                            instance,
+                            promoted: None
+                        };
+                        if let Ok(ct) = cx.tcx.const_eval(cx.param_env.and(global_id)) {
+                            len = ct;
+                        }
+                    }
+
+                    match len.assert_usize(cx.tcx) {
+                        // If the array is definitely non-empty, we can do `#[must_use]` checking.
+                        Some(n) if n != 0 => {
+                            let descr_pre = &format!(
+                                "{}array{} of ",
+                                descr_pre,
+                                plural_suffix,
+                            );
+                            check_must_use_ty(cx, ty, expr, span, descr_pre, descr_post, true)
+                        }
+                        // Otherwise, we don't lint, to avoid false positives.
```

Comment:
I would do this the other way around: always lint unless *definitely* `0`.
Especially if it's generic, that would mean it's *sometimes* `0` (or never, 0-length arrays are rare IMO).